src/vm/contexts.rs

```diff
@@ -63,6 +78,90 @@ pub struct CallStack {
 
 pub type StackTrace = Vec<FunctionIdentifier>;
 
+impl AssetMap {
+    pub fn new() -> AssetMap {
+        AssetMap {
+            map: HashMap::new()
+        }
+    }
+
+    // This will get the next amount for a (principal, asset) entry in the asset table.
+    fn get_next_amount(&self, principal: &PrincipalData, asset: &AssetIdentifier, amount: i128) -> Result<i128> {
+        let current_amount = match self.map.get(principal) {
+            Some(principal_map) => *principal_map.get(&asset).unwrap_or(&0),
+            None => 0
+        };
+            
+        current_amount.checked_add(amount)
+            .ok_or(RuntimeErrorType::ArithmeticOverflow.into())
+    }
+
+    pub fn add_transfer(&mut self, principal: &PrincipalData, asset: AssetIdentifier, amount: i128) -> Result<()> {
```

Comment:
What happens if `amount` is negative?

Reply:
Will update to `panic!` -- since this is only ever called _after_ a transfer has been evaluated and hit the db, it should never be negative.

Reply:
Addressed in 1cf8cf4